<a href="https://colab.research.google.com/github/Ayanlola2002/DATA-SCIENCE-PROJECTS/blob/master/NeedEnergyAPI_Data_Wrangling2withJoe.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [92]:
#mounting Gdrive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [93]:
import pandas as pd
import numpy as np
from datetime import date, datetime
import holidays
import matplotlib.pyplot as plt
from IPython.display import display
pd.options.display.max_columns = None

In [94]:
'''df_5min=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/data_five_min_id_47803.csv')
df_hourly=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/data_hourly_id_47803.csv')
df_daily=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/data_daily_id_47803.csv')
df_monthly=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/data_monthly_id_47803.csv')
df_public_holiday=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/public_holidays_weekends_ZIM.csv')'''


"df_5min=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/data_five_min_id_47803.csv')\ndf_hourly=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/data_hourly_id_47803.csv')\ndf_daily=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/data_daily_id_47803.csv')\ndf_monthly=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/data_monthly_id_47803.csv')\ndf_public_holiday=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/public_holidays_weekends_ZIM.csv')"

In [95]:
##Wrangling Helper Functions to Use
"""
This function do the reading of data file from file path(i.e drive)
"""
def do_read_data(datafile, dateCol):
    ''' Read the excel file'''
    df = pd.read_excel(datafile, index_col= dateCol, parse_dates=True)
    return df

"""
This function Shows Information of the Data file
Like The Header columns, Dimension of Data
"""

def do_data_information(datafile, info = False, head = True, shape = True):
    ''' Look at the info of the data'''
    
    if not isinstance(datafile, pd.DataFrame):
        df = read_data(datafile)
    else:
        df = datafile
    
    ## Information of data
    if info:
        df.info()
        
    print('--------------------------------------------------------------------')
    print('First rows')
    # Show first rows
    if head:
        print(df.head())
        
    print('--------------------------------------------------------------------')
    print('Dimension of the data')
    ## Dimension of data
    if shape:
        print(f'This data has {df.shape[0]} rows and {df.shape[1]} variables') 
        
###This function returns the number of distinct values from any columns       
def do_unique_values(df, col):
    ''' Find the unique values of a columns'''
    return df[col].nunique()


###This function statistically describe the data values
def do_description(df, col = None):
    ''' Give the statistic summary'''
    
    ## Summary statistic
    if col is not None:
        stat = df[col].describe()
    else:
        stat = df.describe()

    ## Change the name of the 50% index to median
    idx = stat.index.tolist()
    idx[5] = 'median'
    stat.index = idx
    return stat

###This function finds % Missing values
def check_missing_value(data):
    ## percentage of missing values
    n = data.isnull().sum().sort_values(ascending=False)/ len(data) * 100
    return n

def check_missing_values_table(df):
  mis_val = df.isnull().sum()
  mis_val_percent = 100 * df.isnull().sum() / len(df)
  mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
  mis_val_table_ren_columns = mis_val_table.rename(
  columns = {0 : 'Missing Values', 1 : '% of Total Values'})
  mis_val_table_ren_columns = mis_val_table_ren_columns[
      mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
  '% of Total Values', ascending=False).round(1)
  print ("Your selected dataframe has " + str(df.shape[1]) + " columns.\n"      
      "There are " + str(mis_val_table_ren_columns.shape[0]) +
        " columns that have missing values.")
  return mis_val_table_ren_columns


###This function finds the amount of each values in a dataset
def do_valueCount(data, col, perc = False):
    ''' calculate the proportion in fraction or percentage'''
    if not perc:
      return data[col].value_counts()
    else:
      return data[col].value_counts(normalize = True).mul(100).round(1).astype(str) + '%'    
### This function count number of unique values for all columns in a dataset
def do_nunique_value(df):
  print('The Numver of unique values per columns are as follows')
  print(df.agg(['nunique']).T)


###For date time conversion start here   
def do_date_col_convertion(df, column_name):
  df[column_name]=pd.to_datetime(df[column_name])
  return df[column_name]

def do_time_col_convertion(df, column_name):
    return pd.to_timedelta(df[column_name])


def do_datepart_extraction(data):
  data['date'] = pd.to_datetime(data['date'])
  data['dateOnly']=data['date'].dt.date
  data['year'] = data['date'].dt.year
  data['quarter']=data['date'].dt.quarter
  data['month'] = data['date'].dt.month
  data['weekday'] = data['date'].dt.weekday
  data['day_name'] = data['date'].dt.day_name()
  data['dayofyear'] = data['date'].dt.dayofyear
  data['day'] =data['date'].dt.day
  data['hour'] = data['date'].dt.hour
  data['minute'] =data['date'].dt.minute
  #data['Season']=data['month'].apply(do_year_season)
  data['season'] = data.date.map(do_season_of_date)
  data['day_section']=data['hour'].apply(do_day_sections)
  return data
   

def do_season_of_date(date):
  year = str(date.year)
  seasons = {'spring': pd.date_range(start='21/03/'+year, end='20/06/'+year),
              'summer': pd.date_range(start='21/06/'+year, end='22/09/'+year),
              'autumn': pd.date_range(start='23/09/'+year, end='20/12/'+year)}
  if date in seasons['spring']:
    return 'spring'
  if date in seasons['summer']:
    return 'summer'
  if date in seasons['autumn']:
    return 'autumn'
  else:
    return 'winter'

# Assuming df has a date column of type `datetime`
#df['season'] = df.date.map(season_of_date)


def do_day_sections(x):
  if (x > 4) and (x <= 8):
    return 'Early Morning'
  elif (x > 8) and (x <= 12 ):
    return 'Morning'
  elif (x > 12) and (x <= 16):
    return'Noon'
  elif (x > 16) and (x <= 20) :
    return 'Evening'
  elif (x > 20) and (x <= 24):
    return'Night'
  elif (x <= 4):
    return'Late Night'




def get_holidays(startYear = 2018, endYear = 2025, countryCode = 'ZA'):
  """
    Takes in a start and end date, and start and end year.
    
    Produces a dataframe with a daily date and columns:
    holiday - 'Y' for holiday
    holidayName - name of the holiday if holiday is 'Y'
    
    Returns a dataframe
  """
  holidayDict = {}
  for i in range(startYear, endYear):
    for date, name in sorted(holidays.CountryHoliday(countryCode,years=[i]).items()):
      holidayDict[date] = name
      holiday_df = pd.DataFrame(list(holidayDict.items()),columns = ['day','holidayName'])
      holiday_df['day'] = pd.to_datetime(holiday_df['day'], format='%Y-%M-%d').dt.date
    return holiday_df
    

    
def get_days(start = '1/1/2018',startYear = 2018, end = '31/12/2025',endYear = 2025, countryCode = 'ZA'):
    """
    Takes in a start and end date, and start and end year.
    
    Produces a dataframe with a daily date and columns:
    weekend - 'Y' for weendend and 'N' for workday
    dayOfweek - numerical day of the week identifier 0 for monday
    weekNum - numerical number of the week 
    holiday - 'Y' for holiday
    holidayName - name of the holiday if holiday is 'Y'
    
    Returns a dataframe
    """
    
    #generate the range of daily dates
    dates = pd.date_range(start = start, end = end)
    date_df = pd.DataFrame(dates, columns = ['day'])
    date_df['day'] = pd.to_datetime(date_df['day'], format='%Y-%M-%d')
    country_holidays = get_holidays(startYear = startYear, endYear = endYear, countryCode = countryCode)
    
    date_df['dayName'] = pd.DatetimeIndex(date_df['day']).day_name()
    date_df['dayOfWeek'] = date_df['day'].dt.dayofweek
    date_df['weekend'] = date_df['dayOfWeek'].apply(lambda x: 'Y' if x>4 else 'N')
    date_df['weekNum'] = date_df['day'].dt.week
    date_df['holiday'] = date_df['day'].apply(lambda x: 'Y' if x in country_holidays['day'].values else 'N')
    date_df['day'] = date_df['day'].dt.date
    date_df = date_df.merge(country_holidays, on='day', how='left', indicator=False)
    
    date_df.to_csv(f'../data/public_holidays_weekends.csv', index=False)
    return date_df

###Functions to impute nan values for numerical varriable
def Impute_nan(df):
  numericalcol = [f for f in df_5min.columns if df_5min.dtypes[f] != 'object']
  missing_col=df_5min[numericalcol].columns[df_5min[numericalcol].isnull().any()].tolist()
  #Technique 1: Using mean to impute the missing values
  for i in missing_col:
    df.loc[df.loc[:,i].isnull(),i]=df.loc[:,i].mean()

def Impute_na(df):
  numericalcol = [f for f in df_5min.columns if df_5min.dtypes[f] != 'object']
  missing_col=df_5min[numericalcol].columns[df_5min[numericalcol].isnull().any()].tolist()
  #Technique 1: Using mean to impute the missing values
  for i in missing_col:
     df[i].fillna(mean,inplace=True)


#dataList=[df_5min,df_hourly,df_daily,df_monthly]

def do_wrangling_task(dL):
  for i in range(len(dL)):
    print(do_data_information(dL[i]))
    print('===============================================')
    print(check_missing_values_table(dL[i]))
    print('===============================================')
    print(do_nunique_value(dL[i]))
    print('===============================================')
    #Feature Engineering
    dL[i]=do_datepart_extraction(dL[i])
    print('===============================================')
    dL[i].date.min(),dL[i].date.max()
    print('===============================================')
    dL[i].consumption.min(),dL[i].consumption.max()
    print('===============================================')
    #checking for days where there weere no consumptions
    dL[i][dL[i].consumption==0].value_counts().count(),dL[i][dL[i].consumption==0].value_counts()
    print('===============================================')
    dL[i]['Public_Holiday'] = dL[i]['dateOnly'].map(mapping)
    print('===============================================')
    dL[i]['Public_Holiday']=dL[i]['Public_Holiday'].apply(lambda x:np.NaN if isinstance(x, str) and (x.isspace() or not x) else x).replace(np.NaN,'Non_Holiday')
    print('===============================================')
    dL[i]=dL[i][['date','timestamp','year','quarter','month','dayofyear','day','day_name','hour','minute','season','day_section','Public_Holiday','consumption','solar','alwaysOn','gridImport','gridExport','selfConsumption','selfSufficiency','active','reactive','voltages','phaseVoltages','currentHarmonics','voltageHarmonics']]
    print('===============================================')
    print('===============================================')

# load helper functions for dynamic import
def import_package(name):
    try:
        module = __import__(name)
    except:
        if name == 'pandas_profiling':
          ! pip install https://github.com/pandas-profiling/pandas-profiling/archive/master.zip
        module = __import__(name)
    try:
        print(f'{name}'+ ' Version:', module.__version__)
    except AttributeError:
        pass
    return module


# JMcQ
def plot_period(df, date1, date2, column, title):
    
    df = df.loc[date1:date2]
    
    fig, axs = plt.subplots(1,1, figsize = (12, 6))
    
    fig.suptitle(title)
    
    x = df['datetime']
    
    axs.plot(x, df[column], label=column)
    axs.legend(loc="best")

    plt.show()

# JMcQ
def create_monthly_profiles(df, column):
    
    months = df['monthName'].unique()
    
    df = df.groupby([df['weekend'], df['monthName'], df['hour']]).mean()
       
    fig, axs = plt.subplots(1, 2, figsize=(10,6), sharey=True, sharex=True)
    
    fig.suptitle('Weekend and Weekday Profiles')
    
    x = list(range(0,24))
    plt.xticks(x)
    
    # Loop through each month of the year and plot profile
    for month in months:
        
        axs[0].plot(x, df.loc[0, month][column], label=month)
        axs[0].legend(loc="best")
        axs[0].set_title('Weekday')

        axs[1].plot(x, df.loc[1, month][column], label=month)
        axs[1].legend(loc="best")
        axs[1].set_title('Weekend')
        
            
    plt.show()
    
    return df

def load_data(file):
    df = pd.read_csv(file)
    df = df.astype({'date':'datetime64[ns]'})
    df.sort_values(by='date', inplace=True)
    return df

#df['holidays']=df['date'].apply(lambda(x: 'Hols' if x in get_holidays()['day'].values else 'Non-Hols'))



In [96]:
asset = "/content/drive/MyDrive/NEEDENERGY/data_five_min_id_47803.csv"

In [97]:
pd = import_package('pandas')
profile = import_package('pandas_profiling')
np = import_package('numpy')
tz = import_package('dateutil').tz
datetime = import_package('datetime').datetime

pandas Version: 1.1.5
numpy Version: 1.19.5
dateutil Version: 2.8.1


In [98]:
df_5min = load_data(asset)

In [99]:
df_5min.head()

date      timestamp  consumption  solar  alwaysOn  \
0 2020-03-15 18:15:00  1584288900000          0.0    0.0       0.0   
1 2020-03-15 18:20:00  1584289200000          0.0    0.0       0.0   
2 2020-03-15 18:25:00  1584289500000          0.0    0.0       0.0   
3 2020-03-15 18:30:00  1584289800000          0.0    0.0       0.0   
4 2020-03-15 18:35:00  1584290100000          0.0    0.0       0.0   

   gridImport  gridExport  selfConsumption  selfSufficiency  \
0         0.0         0.0              0.0              0.0   
1         0.0         0.0              0.0              0.0   
2         0.0         0.0              0.0              0.0   
3         0.0         0.0              0.0              0.0   
4         0.0         0.0              0.0              0.0   

                                              active  \
0  [0.1, 3.7, 0.0, -81.6, -60.0, 30.2, -230.6, -0...   
1  [0.1, 3.8, 0.0, -82.2, -58.5, 30.5, -233.1, -0...   
2  [0.1, 28.5, -26.5, -74.4, -79.2, 55.5, -228.5,...   
3  [0.5, 21.3, -18.6, -76.9, -54.7, 48.7, -227.8,...   
4  [0.1, 3.7, 0.0, -82.6, -33.2, 30.7, 96.6, 0.2,...   

                                            reactive             voltages  \
0  [0.5, 0.9, 0.0, 55.4, 132.4, 14.0, 137.5, 0.2,...  [244.5, None, None]   
1  [0.5, 0.9, 0.0, 55.7, 131.9, 14.3, 144.3, 0.2,...  [246.1, None, None]   
2  [0.5, 19.0, 8.8, 79.6, 137.6, 29.4, 175.0, 0.4...  [246.4, None, None]   
3  [2.2, 13.3, 6.5, 72.4, 135.9, 24.8, 152.8, 0.8...  [246.0, None, None]   
4  [0.5, 0.9, 0.0, 55.2, 128.1, 13.9, 87.7, 0.7, ...  [246.2, None, None]   

         phaseVoltages                                   currentHarmonics  \
0  [244.5, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
1  [246.1, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
2  [246.4, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
3  [246.0, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
4  [246.2, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   

                                    voltageHarmonics  
0  [[], [], [], [], [], [], [], [], [], [], [], [...  
1  [[], [], [], [], [], [], [], [], [], [], [], [...  
2  [[], [], [], [], [], [], [], [], [], [], [], [...  
3  [[], [], [], [], [], [], [], [], [], [], [], [...  
4  [[], [], [], [], [], [], [], [], [], [], [], [...

Data Quality Information:
df_5min : The data has 95649 rows and 15 features

In [100]:
#basic profiling
df_5min.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95649 entries, 0 to 95648
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   date              95649 non-null  datetime64[ns]
 1   timestamp         95649 non-null  int64         
 2   consumption       95649 non-null  float64       
 3   solar             95649 non-null  float64       
 4   alwaysOn          95649 non-null  float64       
 5   gridImport        95649 non-null  float64       
 6   gridExport        95649 non-null  float64       
 7   selfConsumption   95649 non-null  float64       
 8   selfSufficiency   95649 non-null  float64       
 9   active            95649 non-null  object        
 10  reactive          95649 non-null  object        
 11  voltages          95649 non-null  object        
 12  phaseVoltages     95649 non-null  object        
 13  currentHarmonics  95649 non-null  object        
 14  voltageHarmonics  9564

In [101]:
#statistcal summary
df_5min.describe


<bound method NDFrame.describe of                      date      timestamp  consumption   solar  alwaysOn  \
0     2020-03-15 18:15:00  1584288900000        0.000   0.000     0.000   
1     2020-03-15 18:20:00  1584289200000        0.000   0.000     0.000   
2     2020-03-15 18:25:00  1584289500000        0.000   0.000     0.000   
3     2020-03-15 18:30:00  1584289800000        0.000   0.000     0.000   
4     2020-03-15 18:35:00  1584290100000        0.000   0.000     0.000   
...                   ...            ...          ...     ...       ...   
95644 2021-02-12 01:35:00  1613086500000      191.070  41.140  1006.212   
95645 2021-02-12 01:40:00  1613086800000      166.362  47.789  1006.212   
95646 2021-02-12 01:45:00  1613087100000      155.303  41.008  1006.212   
95647 2021-02-12 01:50:00  1613087400000      156.402  48.199  1006.212   
95648 2021-02-12 01:55:00  1613087700000      153.934  46.152  1006.212   

       gridImport  gridExport  selfConsumption  selfSufficiency  

In [102]:
#List of features
df_5min.columns

Index(['date', 'timestamp', 'consumption', 'solar', 'alwaysOn', 'gridImport',
       'gridExport', 'selfConsumption', 'selfSufficiency', 'active',
       'reactive', 'voltages', 'phaseVoltages', 'currentHarmonics',
       'voltageHarmonics'],
      dtype='object')

In [103]:
#profile.ProfileReport(df_5min)
!pip install pandas-profiling[notebook]

In [104]:
#profile.ProfileReport(df_5min)

In [105]:
df_5min_new = df_5min.copy()

In [106]:
df_5min_new.head()

date      timestamp  consumption  solar  alwaysOn  \
0 2020-03-15 18:15:00  1584288900000          0.0    0.0       0.0   
1 2020-03-15 18:20:00  1584289200000          0.0    0.0       0.0   
2 2020-03-15 18:25:00  1584289500000          0.0    0.0       0.0   
3 2020-03-15 18:30:00  1584289800000          0.0    0.0       0.0   
4 2020-03-15 18:35:00  1584290100000          0.0    0.0       0.0   

   gridImport  gridExport  selfConsumption  selfSufficiency  \
0         0.0         0.0              0.0              0.0   
1         0.0         0.0              0.0              0.0   
2         0.0         0.0              0.0              0.0   
3         0.0         0.0              0.0              0.0   
4         0.0         0.0              0.0              0.0   

                                              active  \
0  [0.1, 3.7, 0.0, -81.6, -60.0, 30.2, -230.6, -0...   
1  [0.1, 3.8, 0.0, -82.2, -58.5, 30.5, -233.1, -0...   
2  [0.1, 28.5, -26.5, -74.4, -79.2, 55.5, -228.5,...   
3  [0.5, 21.3, -18.6, -76.9, -54.7, 48.7, -227.8,...   
4  [0.1, 3.7, 0.0, -82.6, -33.2, 30.7, 96.6, 0.2,...   

                                            reactive             voltages  \
0  [0.5, 0.9, 0.0, 55.4, 132.4, 14.0, 137.5, 0.2,...  [244.5, None, None]   
1  [0.5, 0.9, 0.0, 55.7, 131.9, 14.3, 144.3, 0.2,...  [246.1, None, None]   
2  [0.5, 19.0, 8.8, 79.6, 137.6, 29.4, 175.0, 0.4...  [246.4, None, None]   
3  [2.2, 13.3, 6.5, 72.4, 135.9, 24.8, 152.8, 0.8...  [246.0, None, None]   
4  [0.5, 0.9, 0.0, 55.2, 128.1, 13.9, 87.7, 0.7, ...  [246.2, None, None]   

         phaseVoltages                                   currentHarmonics  \
0  [244.5, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
1  [246.1, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
2  [246.4, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
3  [246.0, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
4  [246.2, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   

                                    voltageHarmonics  
0  [[], [], [], [], [], [], [], [], [], [], [], [...  
1  [[], [], [], [], [], [], [], [], [], [], [], [...  
2  [[], [], [], [], [], [], [], [], [], [], [], [...  
3  [[], [], [], [], [], [], [], [], [], [], [], [...  
4  [[], [], [], [], [], [], [], [], [], [], [], [...

In [107]:
#Check for duplicates
df_5min_new.index[df_5min_new['date'].duplicated(keep=False)==True]

Int64Index([], dtype='int64')

In [108]:
# Check for missing info using Missingno Package
msno = import_package('missingno')
msno.matrix(df_5min_new)

missingno Version: 0.4.2


In [109]:
# Check if there are any periods of zero values
numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
features = df_5min_new.select_dtypes(include=numerics).columns 
for feature in features: 
    zero_results = df_5min_new[df_5min_new[feature]==0]
zero_results.head()

date      timestamp  consumption  solar  alwaysOn  \
0 2020-03-15 18:15:00  1584288900000          0.0    0.0       0.0   
1 2020-03-15 18:20:00  1584289200000          0.0    0.0       0.0   
2 2020-03-15 18:25:00  1584289500000          0.0    0.0       0.0   
3 2020-03-15 18:30:00  1584289800000          0.0    0.0       0.0   
4 2020-03-15 18:35:00  1584290100000          0.0    0.0       0.0   

   gridImport  gridExport  selfConsumption  selfSufficiency  \
0         0.0         0.0              0.0              0.0   
1         0.0         0.0              0.0              0.0   
2         0.0         0.0              0.0              0.0   
3         0.0         0.0              0.0              0.0   
4         0.0         0.0              0.0              0.0   

                                              active  \
0  [0.1, 3.7, 0.0, -81.6, -60.0, 30.2, -230.6, -0...   
1  [0.1, 3.8, 0.0, -82.2, -58.5, 30.5, -233.1, -0...   
2  [0.1, 28.5, -26.5, -74.4, -79.2, 55.5, -228.5,...   
3  [0.5, 21.3, -18.6, -76.9, -54.7, 48.7, -227.8,...   
4  [0.1, 3.7, 0.0, -82.6, -33.2, 30.7, 96.6, 0.2,...   

                                            reactive             voltages  \
0  [0.5, 0.9, 0.0, 55.4, 132.4, 14.0, 137.5, 0.2,...  [244.5, None, None]   
1  [0.5, 0.9, 0.0, 55.7, 131.9, 14.3, 144.3, 0.2,...  [246.1, None, None]   
2  [0.5, 19.0, 8.8, 79.6, 137.6, 29.4, 175.0, 0.4...  [246.4, None, None]   
3  [2.2, 13.3, 6.5, 72.4, 135.9, 24.8, 152.8, 0.8...  [246.0, None, None]   
4  [0.5, 0.9, 0.0, 55.2, 128.1, 13.9, 87.7, 0.7, ...  [246.2, None, None]   

         phaseVoltages                                   currentHarmonics  \
0  [244.5, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
1  [246.1, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
2  [246.4, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
3  [246.0, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
4  [246.2, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   

                                    voltageHarmonics  
0  [[], [], [], [], [], [], [], [], [], [], [], [...  
1  [[], [], [], [], [], [], [], [], [], [], [], [...  
2  [[], [], [], [], [], [], [], [], [], [], [], [...  
3  [[], [], [], [], [], [], [], [], [], [], [], [...  
4  [[], [], [], [], [], [], [], [], [], [], [], [...

In [110]:
# Check for any missing time steps, create hourly datetime index and compare to length of dataframe
start = df_5min_new['date'].min()
end = df_5min_new['date'].max()
full_index = pd.date_range(start, end, freq='H')
len(full_index) == len(df_5min_new)

False

In [111]:
df_5min_new.set_index('date', inplace = True)
df_5min_new = df_5min_new.reindex(full_index)
df_5min_new.index

DatetimeIndex(['2020-03-15 18:15:00', '2020-03-15 19:15:00',
               '2020-03-15 20:15:00', '2020-03-15 21:15:00',
               '2020-03-15 22:15:00', '2020-03-15 23:15:00',
               '2020-03-16 00:15:00', '2020-03-16 01:15:00',
               '2020-03-16 02:15:00', '2020-03-16 03:15:00',
               ...
               '2021-02-11 16:15:00', '2021-02-11 17:15:00',
               '2021-02-11 18:15:00', '2021-02-11 19:15:00',
               '2021-02-11 20:15:00', '2021-02-11 21:15:00',
               '2021-02-11 22:15:00', '2021-02-11 23:15:00',
               '2021-02-12 00:15:00', '2021-02-12 01:15:00'],
              dtype='datetime64[ns]', length=8000, freq='H')

In [112]:
#Have a look at missing values
print(f"{df_5min_new.isnull().sum()}\n\n")
print(f"Percentage of missing values:\t{(df_5min_new['consumption'].isnull().sum()/df_5min_new['timestamp'].count())*100}")

timestamp           28
consumption         28
solar               28
alwaysOn            28
gridImport          28
gridExport          28
selfConsumption     28
selfSufficiency     28
active              28
reactive            28
voltages            28
phaseVoltages       28
currentHarmonics    28
voltageHarmonics    28
dtype: int64


Percentage of missing values:	0.35122930255895635


Feature Engineering Time Series

In [113]:
df_5min_new.reset_index(inplace=True)
df_5min_new.rename(columns = {'index':'datetime'}, inplace = True)
df_5min_new.head()

datetime     timestamp  consumption  solar  alwaysOn  gridImport  \
0 2020-03-15 18:15:00  1.584289e+12        0.000    0.0       0.0       0.000   
1 2020-03-15 19:15:00  1.584292e+12       35.615    0.0       0.0      35.615   
2 2020-03-15 20:15:00  1.584296e+12       49.976    0.0       0.0      49.976   
3 2020-03-15 21:15:00  1.584300e+12       67.103    0.0       0.0      67.103   
4 2020-03-15 22:15:00  1.584303e+12       49.070    0.0       0.0      49.070   

   gridExport  selfConsumption  selfSufficiency  \
0         0.0              0.0              0.0   
1         0.0              0.0              0.0   
2         0.0              0.0              0.0   
3         0.0              0.0              0.0   
4         0.0              0.0              0.0   

                                              active  \
0  [0.1, 3.7, 0.0, -81.6, -60.0, 30.2, -230.6, -0...   
1  [0.1, 20.3, 0.0, None, 0.0, 47.0, 237.2, 0.3, ...   
2  [0.1, 3.8, 0.0, None, None, 33.4, 117.8, 0.4, ...   
3  [0.1, 3.7, 0.0, None, None, 27.4, 92.3, 0.4, N...   
4  [0.1, 3.7, 0.0, None, None, 27.3, 91.4, 0.4, N...   

                                            reactive             voltages  \
0  [0.5, 0.9, 0.0, 55.4, 132.4, 14.0, 137.5, 0.2,...  [244.5, None, None]   
1  [0.5, 1.9, 0.0, None, 0.1, 13.6, 155.9, 0.2, N...  [247.3, None, None]   
2  [0.5, 0.9, 0.0, None, None, 14.2, 147.8, 0.3, ...  [246.7, None, None]   
3  [0.5, 0.9, 0.0, None, None, 12.1, 113.0, 0.3, ...  [244.5, None, None]   
4  [0.5, 0.9, 0.0, None, None, 12.6, 113.2, 0.3, ...  [243.8, None, None]   

         phaseVoltages                                   currentHarmonics  \
0  [244.5, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
1  [247.3, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
2  [246.7, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
3  [244.5, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
4  [243.8, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   

                                    voltageHarmonics  
0  [[], [], [], [], [], [], [], [], [], [], [], [...  
1  [[], [], [], [], [], [], [], [], [], [], [], [...  
2  [[], [], [], [], [], [], [], [], [], [], [], [...  
3  [[], [], [], [], [], [], [], [], [], [], [], [...  
4  [[], [], [], [], [], [], [], [], [], [], [], [...

In [114]:
#Feature Engineering of Time Series
#df_5min_new['datetime']=pd.to_datetime(df_5min_new['date'])
df_5min_new['date'] = df_5min_new['datetime'].dt.strftime('%Y-%m-%d')
df_5min_new['year'] = df_5min_new['datetime'].dt.year 
df_5min_new['dayOfYear'] = df_5min_new['datetime'].dt.dayofyear
df_5min_new['month'] = df_5min_new['datetime'].dt.month 
df_5min_new['monthName'] = pd.DatetimeIndex(df_5min_new['datetime']).month_name()
df_5min_new['week'] = df_5min_new['datetime'].dt.week
df_5min_new['day'] = df_5min_new['datetime'].dt.day
df_5min_new['dayName'] = pd.DatetimeIndex(df_5min_new['datetime']).day_name()
df_5min_new['hour'] = df_5min_new['datetime'].dt.hour
df_5min_new['minute'] = df_5min_new['datetime'].dt.minute
df_5min_new['dayOfWeek'] = df_5min_new['datetime'].dt.dayofweek
df_5min_new['weekend'] = df_5min_new['dayOfWeek'].apply(lambda x: 1 if x >= 5 else 0)
df_5min_new['time'] = df_5min_new['datetime'].dt.time 
df_5min_new['time_Str'] = df_5min_new['datetime'].dt.time.astype(str)   
df_5min_new['dayMonth'] = df_5min_new["datetime"].dt.strftime('%m-%d')
df_5min_new['hourMinute'] = df_5min_new["datetime"].dt.strftime('%H:%M')

bins = [0,4,8,12,16,20,24]
labels = ['Late Night', 'Early Morning','Morning','Noon','Eve','Night']
df_5min_new['session'] = pd.cut(df_5min_new['hour'], bins=bins, labels=labels, include_lowest=True)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated.  Please use Series.dt.isocalendar().week instead.
  


Remove Duplicates

In [115]:
df_5min_new = df_5min_new[~df_5min_new['datetime'].duplicated(keep = 'first')]
df_5min_new.head()

datetime     timestamp  consumption  solar  alwaysOn  gridImport  \
0 2020-03-15 18:15:00  1.584289e+12        0.000    0.0       0.0       0.000   
1 2020-03-15 19:15:00  1.584292e+12       35.615    0.0       0.0      35.615   
2 2020-03-15 20:15:00  1.584296e+12       49.976    0.0       0.0      49.976   
3 2020-03-15 21:15:00  1.584300e+12       67.103    0.0       0.0      67.103   
4 2020-03-15 22:15:00  1.584303e+12       49.070    0.0       0.0      49.070   

   gridExport  selfConsumption  selfSufficiency  \
0         0.0              0.0              0.0   
1         0.0              0.0              0.0   
2         0.0              0.0              0.0   
3         0.0              0.0              0.0   
4         0.0              0.0              0.0   

                                              active  \
0  [0.1, 3.7, 0.0, -81.6, -60.0, 30.2, -230.6, -0...   
1  [0.1, 20.3, 0.0, None, 0.0, 47.0, 237.2, 0.3, ...   
2  [0.1, 3.8, 0.0, None, None, 33.4, 117.8, 0.4, ...   
3  [0.1, 3.7, 0.0, None, None, 27.4, 92.3, 0.4, N...   
4  [0.1, 3.7, 0.0, None, None, 27.3, 91.4, 0.4, N...   

                                            reactive             voltages  \
0  [0.5, 0.9, 0.0, 55.4, 132.4, 14.0, 137.5, 0.2,...  [244.5, None, None]   
1  [0.5, 1.9, 0.0, None, 0.1, 13.6, 155.9, 0.2, N...  [247.3, None, None]   
2  [0.5, 0.9, 0.0, None, None, 14.2, 147.8, 0.3, ...  [246.7, None, None]   
3  [0.5, 0.9, 0.0, None, None, 12.1, 113.0, 0.3, ...  [244.5, None, None]   
4  [0.5, 0.9, 0.0, None, None, 12.6, 113.2, 0.3, ...  [243.8, None, None]   

         phaseVoltages                                   currentHarmonics  \
0  [244.5, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
1  [247.3, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
2  [246.7, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
3  [244.5, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
4  [243.8, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   

                                    voltageHarmonics        date  year  \
0  [[], [], [], [], [], [], [], [], [], [], [], [...  2020-03-15  2020   
1  [[], [], [], [], [], [], [], [], [], [], [], [...  2020-03-15  2020   
2  [[], [], [], [], [], [], [], [], [], [], [], [...  2020-03-15  2020   
3  [[], [], [], [], [], [], [], [], [], [], [], [...  2020-03-15  2020   
4  [[], [], [], [], [], [], [], [], [], [], [], [...  2020-03-15  2020   

   dayOfYear  month monthName  week  day dayName  hour  minute  dayOfWeek  \
0         75      3     March    11   15  Sunday    18      15          6   
1         75      3     March    11   15  Sunday    19      15          6   
2         75      3     March    11   15  Sunday    20      15          6   
3         75      3     March    11   15  Sunday    21      15          6   
4         75      3     March    11   15  Sunday    22      15          6   

   weekend      time  time_Str dayMonth hourMinute session  
0        1  18:15:00  18:15:00    03-15      18:15     Eve  
1        1  19:15:00  19:15:00    03-15      19:15     Eve  
2        1  20:15:00  20:15:00    03-15      20:15     Eve  
3        1  21:15:00  21:15:00    03-15      21:15   Night  
4        1  22:15:00  22:15:00    03-15      22:15   Night

In [116]:
df_5min_new.columns


Index(['datetime', 'timestamp', 'consumption', 'solar', 'alwaysOn',
       'gridImport', 'gridExport', 'selfConsumption', 'selfSufficiency',
       'active', 'reactive', 'voltages', 'phaseVoltages', 'currentHarmonics',
       'voltageHarmonics', 'date', 'year', 'dayOfYear', 'month', 'monthName',
       'week', 'day', 'dayName', 'hour', 'minute', 'dayOfWeek', 'weekend',
       'time', 'time_Str', 'dayMonth', 'hourMinute', 'session'],
      dtype='object')

In [117]:
#Reorder columns
# Put non-ML friendly columns first - they may get dropped when loading into models
col_order = ['datetime', 'consumption_kWh', 'date', 'year', 'dayMonth', 'time', 'time_Str', 'hourMinute', 'session', 'monthName', 'dayName', 
             'dayOfYear', 'week', 'month', 'day', 'hour', 'minute', 'dayOfWeek', 'weekend', 'workingDay']
df_5min_new =df_5min_new.reindex(columns=col_order)

1 hour data wrangling

In [123]:
asset ="/content/drive/MyDrive/NEEDENERGY/data_hourly_id_47803.csv"


In [124]:
df_hourly = load_data(asset)

In [125]:
df_hourly.head()

date      timestamp  consumption  solar  alwaysOn  \
0 2020-03-15 18:00:00  1584288000000        0.000    0.0       0.0   
1 2020-03-15 19:00:00  1584291600000      913.000    0.0       0.0   
2 2020-03-15 20:00:00  1584295200000      623.471    0.0       0.0   
3 2020-03-15 21:00:00  1584298800000      765.984    0.0       0.0   
4 2020-03-15 22:00:00  1584302400000      585.041    0.0       0.0   

   gridImport  gridExport  selfConsumption  selfSufficiency  \
0       0.000         0.0              0.0              0.0   
1     913.000         0.0              0.0              0.0   
2     623.471         0.0              0.0              0.0   
3     765.984         0.0              0.0              0.0   
4     585.041         0.0              0.0              0.0   

                                              active  \
0  [1.3, 78.9, -45.1, -297.3, -163.6, 321.1, 143....   
1  [1.2, 596.1, -509.8, 72.6, 0.0, 907.0, 2273.7,...   
2  [1.5, 44.8, 0.0, None, None, 387.2, 1148.0, 6....   
3  [1.2, 44.4, 0.0, None, None, 332.9, 1118.7, 4....   
4  [1.5, 44.4, 0.0, None, None, 329.0, 1414.6, 5....   

                                            reactive             voltages  \
0  [6.2, 38.9, 15.3, 534.8, 1134.3, 151.0, 1309.9...  [245.8, None, None]   
1  [6.0, 343.5, 228.7, 262.0, 0.2, 439.9, 2077.9,...  [245.6, None, None]   
2  [7.4, 10.8, 0.0, None, None, 155.4, 1374.2, 3....  [244.9, None, None]   
3  [6.0, 10.8, 0.0, None, None, 145.8, 1314.5, 2....  [244.4, None, None]   
4  [7.5, 10.8, 0.0, None, None, 155.1, 1395.1, 3....  [244.8, None, None]   

         phaseVoltages                                   currentHarmonics  \
0  [245.8, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
1  [245.6, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
2  [244.9, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
3  [244.4, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
4  [244.8, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   

                                    voltageHarmonics  
0  [[], [], [], [], [], [], [], [], [], [], [], [...  
1  [[], [], [], [], [], [], [], [], [], [], [], [...  
2  [[], [], [], [], [], [], [], [], [], [], [], [...  
3  [[], [], [], [], [], [], [], [], [], [], [], [...  
4  [[], [], [], [], [], [], [], [], [], [], [], [...

In [126]:
#basic profiling
df_hourly.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7979 entries, 0 to 7978
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   date              7979 non-null   datetime64[ns]
 1   timestamp         7979 non-null   int64         
 2   consumption       7979 non-null   float64       
 3   solar             7979 non-null   float64       
 4   alwaysOn          7979 non-null   float64       
 5   gridImport        7979 non-null   float64       
 6   gridExport        7979 non-null   float64       
 7   selfConsumption   7979 non-null   float64       
 8   selfSufficiency   7979 non-null   float64       
 9   active            7979 non-null   object        
 10  reactive          7979 non-null   object        
 11  voltages          7979 non-null   object        
 12  phaseVoltages     7979 non-null   object        
 13  currentHarmonics  7979 non-null   object        
 14  voltageHarmonics  7979 n

In [127]:
#statistcal summary
df_hourly.describe


<bound method NDFrame.describe of                     date      timestamp  consumption    solar  alwaysOn  \
0    2020-03-15 18:00:00  1584288000000        0.000    0.000     0.000   
1    2020-03-15 19:00:00  1584291600000      913.000    0.000     0.000   
2    2020-03-15 20:00:00  1584295200000      623.471    0.000     0.000   
3    2020-03-15 21:00:00  1584298800000      765.984    0.000     0.000   
4    2020-03-15 22:00:00  1584302400000      585.041    0.000     0.000   
...                  ...            ...          ...      ...       ...   
7974 2021-02-11 21:00:00  1613070000000     1973.376  572.885  1006.751   
7975 2021-02-11 22:00:00  1613073600000     1995.612  576.278  1006.751   
7976 2021-02-11 23:00:00  1613077200000     2020.629  564.869  1006.751   
7977 2021-02-12 00:00:00  1613080800000     1960.142  556.150  1006.751   
7978 2021-02-12 01:00:00  1613084400000     1970.954  555.986  1006.751   

      gridImport  gridExport  selfConsumption  selfSufficiency  \

In [128]:
#List of features
df_hourly.columns

Index(['date', 'timestamp', 'consumption', 'solar', 'alwaysOn', 'gridImport',
       'gridExport', 'selfConsumption', 'selfSufficiency', 'active',
       'reactive', 'voltages', 'phaseVoltages', 'currentHarmonics',
       'voltageHarmonics'],
      dtype='object')

In [129]:
df_hourly_new = df_hourly.copy()

df_hourly_new.head()

date      timestamp  consumption  solar  alwaysOn  \
0 2020-03-15 18:00:00  1584288000000        0.000    0.0       0.0   
1 2020-03-15 19:00:00  1584291600000      913.000    0.0       0.0   
2 2020-03-15 20:00:00  1584295200000      623.471    0.0       0.0   
3 2020-03-15 21:00:00  1584298800000      765.984    0.0       0.0   
4 2020-03-15 22:00:00  1584302400000      585.041    0.0       0.0   

   gridImport  gridExport  selfConsumption  selfSufficiency  \
0       0.000         0.0              0.0              0.0   
1     913.000         0.0              0.0              0.0   
2     623.471         0.0              0.0              0.0   
3     765.984         0.0              0.0              0.0   
4     585.041         0.0              0.0              0.0   

                                              active  \
0  [1.3, 78.9, -45.1, -297.3, -163.6, 321.1, 143....   
1  [1.2, 596.1, -509.8, 72.6, 0.0, 907.0, 2273.7,...   
2  [1.5, 44.8, 0.0, None, None, 387.2, 1148.0, 6....   
3  [1.2, 44.4, 0.0, None, None, 332.9, 1118.7, 4....   
4  [1.5, 44.4, 0.0, None, None, 329.0, 1414.6, 5....   

                                            reactive             voltages  \
0  [6.2, 38.9, 15.3, 534.8, 1134.3, 151.0, 1309.9...  [245.8, None, None]   
1  [6.0, 343.5, 228.7, 262.0, 0.2, 439.9, 2077.9,...  [245.6, None, None]   
2  [7.4, 10.8, 0.0, None, None, 155.4, 1374.2, 3....  [244.9, None, None]   
3  [6.0, 10.8, 0.0, None, None, 145.8, 1314.5, 2....  [244.4, None, None]   
4  [7.5, 10.8, 0.0, None, None, 155.1, 1395.1, 3....  [244.8, None, None]   

         phaseVoltages                                   currentHarmonics  \
0  [245.8, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
1  [245.6, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
2  [244.9, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
3  [244.4, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
4  [244.8, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   

                                    voltageHarmonics  
0  [[], [], [], [], [], [], [], [], [], [], [], [...  
1  [[], [], [], [], [], [], [], [], [], [], [], [...  
2  [[], [], [], [], [], [], [], [], [], [], [], [...  
3  [[], [], [], [], [], [], [], [], [], [], [], [...  
4  [[], [], [], [], [], [], [], [], [], [], [], [...

In [130]:
#Check for duplicates
df_hourly_new.index[df_hourly_new['date'].duplicated(keep=False)==True]

Int64Index([], dtype='int64')

In [131]:
# Check for missing info using Missingno Package
msno = import_package('missingno')
msno.matrix(df_hourly_new)

missingno Version: 0.4.2


In [132]:
# Check if there are any periods of zero values
numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
features = df_hourly_new.select_dtypes(include=numerics).columns 
for feature in features: 
    zero_results = df_hourly_new[df_hourly_new[feature]==0]
zero_results.head()

date      timestamp  consumption  solar  alwaysOn  \
0 2020-03-15 18:00:00  1584288000000        0.000    0.0       0.0   
1 2020-03-15 19:00:00  1584291600000      913.000    0.0       0.0   
2 2020-03-15 20:00:00  1584295200000      623.471    0.0       0.0   
3 2020-03-15 21:00:00  1584298800000      765.984    0.0       0.0   
4 2020-03-15 22:00:00  1584302400000      585.041    0.0       0.0   

   gridImport  gridExport  selfConsumption  selfSufficiency  \
0       0.000         0.0              0.0              0.0   
1     913.000         0.0              0.0              0.0   
2     623.471         0.0              0.0              0.0   
3     765.984         0.0              0.0              0.0   
4     585.041         0.0              0.0              0.0   

                                              active  \
0  [1.3, 78.9, -45.1, -297.3, -163.6, 321.1, 143....   
1  [1.2, 596.1, -509.8, 72.6, 0.0, 907.0, 2273.7,...   
2  [1.5, 44.8, 0.0, None, None, 387.2, 1148.0, 6....   
3  [1.2, 44.4, 0.0, None, None, 332.9, 1118.7, 4....   
4  [1.5, 44.4, 0.0, None, None, 329.0, 1414.6, 5....   

                                            reactive             voltages  \
0  [6.2, 38.9, 15.3, 534.8, 1134.3, 151.0, 1309.9...  [245.8, None, None]   
1  [6.0, 343.5, 228.7, 262.0, 0.2, 439.9, 2077.9,...  [245.6, None, None]   
2  [7.4, 10.8, 0.0, None, None, 155.4, 1374.2, 3....  [244.9, None, None]   
3  [6.0, 10.8, 0.0, None, None, 145.8, 1314.5, 2....  [244.4, None, None]   
4  [7.5, 10.8, 0.0, None, None, 155.1, 1395.1, 3....  [244.8, None, None]   

         phaseVoltages                                   currentHarmonics  \
0  [245.8, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
1  [245.6, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
2  [244.9, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
3  [244.4, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
4  [244.8, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   

                                    voltageHarmonics  
0  [[], [], [], [], [], [], [], [], [], [], [], [...  
1  [[], [], [], [], [], [], [], [], [], [], [], [...  
2  [[], [], [], [], [], [], [], [], [], [], [], [...  
3  [[], [], [], [], [], [], [], [], [], [], [], [...  
4  [[], [], [], [], [], [], [], [], [], [], [], [...

In [133]:
# Check for any missing time steps, create hourly datetime index and compare to length of dataframe
start = df_hourly_new['date'].min()
end = df_hourly_new['date'].max()
full_index = pd.date_range(start, end, freq='H')
len(full_index) == len(df_hourly_new)

False

In [134]:
df_hourly_new.set_index('date', inplace = True)
df_hourly_new = df_hourly_new.reindex(full_index)
df_hourly_new.index

DatetimeIndex(['2020-03-15 18:00:00', '2020-03-15 19:00:00',
               '2020-03-15 20:00:00', '2020-03-15 21:00:00',
               '2020-03-15 22:00:00', '2020-03-15 23:00:00',
               '2020-03-16 00:00:00', '2020-03-16 01:00:00',
               '2020-03-16 02:00:00', '2020-03-16 03:00:00',
               ...
               '2021-02-11 16:00:00', '2021-02-11 17:00:00',
               '2021-02-11 18:00:00', '2021-02-11 19:00:00',
               '2021-02-11 20:00:00', '2021-02-11 21:00:00',
               '2021-02-11 22:00:00', '2021-02-11 23:00:00',
               '2021-02-12 00:00:00', '2021-02-12 01:00:00'],
              dtype='datetime64[ns]', length=8000, freq='H')

In [135]:
#Have a look at missing values
print(f"{df_hourly_new.isnull().sum()}\n\n")
print(f"Percentage of missing values:\t{(df_hourly_new['consumption'].isnull().sum()/df_hourly_new['timestamp'].count())*100}")

timestamp           21
consumption         21
solar               21
alwaysOn            21
gridImport          21
gridExport          21
selfConsumption     21
selfSufficiency     21
active              21
reactive            21
voltages            21
phaseVoltages       21
currentHarmonics    21
voltageHarmonics    21
dtype: int64


Percentage of missing values:	0.26319087604963026


In [136]:

df_hourly_new.reset_index(inplace=True)
df_hourly_new.rename(columns = {'index':'datetime'}, inplace = True)
df_hourly_new.head()

datetime     timestamp  consumption  solar  alwaysOn  gridImport  \
0 2020-03-15 18:00:00  1.584288e+12        0.000    0.0       0.0       0.000   
1 2020-03-15 19:00:00  1.584292e+12      913.000    0.0       0.0     913.000   
2 2020-03-15 20:00:00  1.584295e+12      623.471    0.0       0.0     623.471   
3 2020-03-15 21:00:00  1.584299e+12      765.984    0.0       0.0     765.984   
4 2020-03-15 22:00:00  1.584302e+12      585.041    0.0       0.0     585.041   

   gridExport  selfConsumption  selfSufficiency  \
0         0.0              0.0              0.0   
1         0.0              0.0              0.0   
2         0.0              0.0              0.0   
3         0.0              0.0              0.0   
4         0.0              0.0              0.0   

                                              active  \
0  [1.3, 78.9, -45.1, -297.3, -163.6, 321.1, 143....   
1  [1.2, 596.1, -509.8, 72.6, 0.0, 907.0, 2273.7,...   
2  [1.5, 44.8, 0.0, None, None, 387.2, 1148.0, 6....   
3  [1.2, 44.4, 0.0, None, None, 332.9, 1118.7, 4....   
4  [1.5, 44.4, 0.0, None, None, 329.0, 1414.6, 5....   

                                            reactive             voltages  \
0  [6.2, 38.9, 15.3, 534.8, 1134.3, 151.0, 1309.9...  [245.8, None, None]   
1  [6.0, 343.5, 228.7, 262.0, 0.2, 439.9, 2077.9,...  [245.6, None, None]   
2  [7.4, 10.8, 0.0, None, None, 155.4, 1374.2, 3....  [244.9, None, None]   
3  [6.0, 10.8, 0.0, None, None, 145.8, 1314.5, 2....  [244.4, None, None]   
4  [7.5, 10.8, 0.0, None, None, 155.1, 1395.1, 3....  [244.8, None, None]   

         phaseVoltages                                   currentHarmonics  \
0  [245.8, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
1  [245.6, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
2  [244.9, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
3  [244.4, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
4  [244.8, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   

                                    voltageHarmonics  
0  [[], [], [], [], [], [], [], [], [], [], [], [...  
1  [[], [], [], [], [], [], [], [], [], [], [], [...  
2  [[], [], [], [], [], [], [], [], [], [], [], [...  
3  [[], [], [], [], [], [], [], [], [], [], [], [...  
4  [[], [], [], [], [], [], [], [], [], [], [], [...

In [137]:
#Feature Engineering of Time Series
#df_5min_new['datetime']=pd.to_datetime(df_5min_new['date'])
df_hourly_new['date'] = df_hourly_new['datetime'].dt.strftime('%Y-%m-%d')
df_hourly_new['year'] = df_hourly_new['datetime'].dt.year 
df_hourly_new['dayOfYear'] = df_hourly_new['datetime'].dt.dayofyear
df_hourly_new['month'] = df_hourly_new['datetime'].dt.month 
df_hourly_new['monthName'] = pd.DatetimeIndex(df_hourly_new['datetime']).month_name()
df_hourly_new['week'] = df_hourly_new['datetime'].dt.week
df_hourly_new['day'] = df_hourly_new['datetime'].dt.day
df_hourly_new['dayName'] = pd.DatetimeIndex(df_hourly_new['datetime']).day_name()
df_hourly_new['hour'] = df_hourly_new['datetime'].dt.hour
df_hourly_new['minute'] = df_hourly_new['datetime'].dt.minute
df_hourly_new['dayOfWeek'] = df_hourly_new['datetime'].dt.dayofweek
df_hourly_new['weekend'] = df_hourly_new['dayOfWeek'].apply(lambda x: 1 if x >= 5 else 0)
df_hourly_new['time'] = df_hourly_new['datetime'].dt.time 
df_hourly_new['time_Str'] = df_hourly_new['datetime'].dt.time.astype(str)   
df_hourly_new['dayMonth'] = df_hourly_new["datetime"].dt.strftime('%m-%d')
df_hourly_new['hourMinute'] = df_hourly_new["datetime"].dt.strftime('%H:%M')

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated.  Please use Series.dt.isocalendar().week instead.
  


In [138]:
bins = [0,4,8,12,16,20,24]
labels = ['Late Night', 'Early Morning','Morning','Noon','Eve','Night']
df_hourly_new['session'] = pd.cut(df_hourly_new['hour'], bins=bins, labels=labels, include_lowest=True)

In [139]:
df_hourly_new = df_hourly_new[~df_hourly_new['datetime'].duplicated(keep = 'first')]
df_hourly_new.head()


datetime     timestamp  consumption  solar  alwaysOn  gridImport  \
0 2020-03-15 18:00:00  1.584288e+12        0.000    0.0       0.0       0.000   
1 2020-03-15 19:00:00  1.584292e+12      913.000    0.0       0.0     913.000   
2 2020-03-15 20:00:00  1.584295e+12      623.471    0.0       0.0     623.471   
3 2020-03-15 21:00:00  1.584299e+12      765.984    0.0       0.0     765.984   
4 2020-03-15 22:00:00  1.584302e+12      585.041    0.0       0.0     585.041   

   gridExport  selfConsumption  selfSufficiency  \
0         0.0              0.0              0.0   
1         0.0              0.0              0.0   
2         0.0              0.0              0.0   
3         0.0              0.0              0.0   
4         0.0              0.0              0.0   

                                              active  \
0  [1.3, 78.9, -45.1, -297.3, -163.6, 321.1, 143....   
1  [1.2, 596.1, -509.8, 72.6, 0.0, 907.0, 2273.7,...   
2  [1.5, 44.8, 0.0, None, None, 387.2, 1148.0, 6....   
3  [1.2, 44.4, 0.0, None, None, 332.9, 1118.7, 4....   
4  [1.5, 44.4, 0.0, None, None, 329.0, 1414.6, 5....   

                                            reactive             voltages  \
0  [6.2, 38.9, 15.3, 534.8, 1134.3, 151.0, 1309.9...  [245.8, None, None]   
1  [6.0, 343.5, 228.7, 262.0, 0.2, 439.9, 2077.9,...  [245.6, None, None]   
2  [7.4, 10.8, 0.0, None, None, 155.4, 1374.2, 3....  [244.9, None, None]   
3  [6.0, 10.8, 0.0, None, None, 145.8, 1314.5, 2....  [244.4, None, None]   
4  [7.5, 10.8, 0.0, None, None, 155.1, 1395.1, 3....  [244.8, None, None]   

         phaseVoltages                                   currentHarmonics  \
0  [245.8, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
1  [245.6, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
2  [244.9, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
3  [244.4, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   
4  [244.8, None, None]  [[], [], [], [], [], [], [], [], [], [], [], [...   

                                    voltageHarmonics        date  year  \
0  [[], [], [], [], [], [], [], [], [], [], [], [...  2020-03-15  2020   
1  [[], [], [], [], [], [], [], [], [], [], [], [...  2020-03-15  2020   
2  [[], [], [], [], [], [], [], [], [], [], [], [...  2020-03-15  2020   
3  [[], [], [], [], [], [], [], [], [], [], [], [...  2020-03-15  2020   
4  [[], [], [], [], [], [], [], [], [], [], [], [...  2020-03-15  2020   

   dayOfYear  month monthName  week  day dayName  hour  minute  dayOfWeek  \
0         75      3     March    11   15  Sunday    18       0          6   
1         75      3     March    11   15  Sunday    19       0          6   
2         75      3     March    11   15  Sunday    20       0          6   
3         75      3     March    11   15  Sunday    21       0          6   
4         75      3     March    11   15  Sunday    22       0          6   

   weekend      time  time_Str dayMonth hourMinute session  
0        1  18:00:00  18:00:00    03-15      18:00     Eve  
1        1  19:00:00  19:00:00    03-15      19:00     Eve  
2        1  20:00:00  20:00:00    03-15      20:00     Eve  
3        1  21:00:00  21:00:00    03-15      21:00   Night  
4        1  22:00:00  22:00:00    03-15      22:00   Night

In [140]:
df_hourly_new.columns

Index(['datetime', 'timestamp', 'consumption', 'solar', 'alwaysOn',
       'gridImport', 'gridExport', 'selfConsumption', 'selfSufficiency',
       'active', 'reactive', 'voltages', 'phaseVoltages', 'currentHarmonics',
       'voltageHarmonics', 'date', 'year', 'dayOfYear', 'month', 'monthName',
       'week', 'day', 'dayName', 'hour', 'minute', 'dayOfWeek', 'weekend',
       'time', 'time_Str', 'dayMonth', 'hourMinute', 'session'],
      dtype='object')

In [141]:
#Reorder columns
# Put non-ML friendly columns first - they may get dropped when loading into models
col_order = ['datetime', 'consumption_kWh', 'date', 'year', 'dayMonth', 'time', 'time_Str', 'hourMinute', 'session', 'monthName', 'dayName', 
             'dayOfYear', 'week', 'month', 'day', 'hour', 'minute', 'dayOfWeek', 'weekend', 'workingDay']
df_hourly_new =df_hourly_new.reindex(columns=col_order)

daily Wrangling

In [ ]:
asset = "/content/drive/MyDrive/NEEDENERGY/data_daily_id_47803.csv"


In [142]:
df_daily = load_data(asset)

In [144]:
#df_daily.head()

In [ ]:
#basic profiling
df_daily.info()

In [ ]:
#statistcal summary
df_daily.describe

In [ ]:
df_daily_new = df_daily.copy()

In [ ]:
df_daily_new.head()

In [ ]:
#Check for duplicates
df_daily_new.index[df_daily_new['date'].duplicated(keep=False)==True]

In [ ]:
# Check for missing info using Missingno Package
msno = import_package('missingno')
msno.matrix(df_daily_new)

In [ ]:
# Check if there are any periods of zero values
numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
features = df_daily_new.select_dtypes(include=numerics).columns 
for feature in features: 
    zero_results = df_daily_new[df_daily_new[feature]==0]
zero_results.head()

In [ ]:
# Check for any missing time steps, create hourly datetime index and compare to length of dataframe
start = df_daily_new['date'].min()
end = df_daily_new['date'].max()
full_index = pd.date_range(start, end, freq='H')
len(full_index) == len(df_daily_new)

In [ ]:

df_daily_new.set_index('date', inplace = True)
df_daily_new = df_daily_new.reindex(full_index)
df_daily_new.index

In [ ]:
#Have a look at missing values
print(f"{df_daily_new.isnull().sum()}\n\n")
print(f"Percentage of missing values:\t{(df_daily_new['consumption'].isnull().sum()/df_daily_new['timestamp'].count())*100}")

In [ ]:
df_daily_new.reset_index(inplace=True)
df_daily_new.rename(columns = {'index':'datetime'}, inplace = True)
df_daily_new.head()

In [ ]:
#Feature Engineering of Time Series
#df_5min_new['datetime']=pd.to_datetime(df_5min_new['date'])
df_daily_new['date'] = df_daily_new['datetime'].dt.strftime('%Y-%m-%d')
df_daily_new['year'] = df_daily_new['datetime'].dt.year 
df_daily_new['dayOfYear'] = df_daily_new['datetime'].dt.dayofyear
df_daily_new['month'] = df_daily_new['datetime'].dt.month 
df_daily_new['monthName'] = pd.DatetimeIndex(df_daily_new['datetime']).month_name()
df_daily_new['week'] = df_daily_new['datetime'].dt.week
df_daily_new['day'] = df_daily_new['datetime'].dt.day
df_daily_new['dayName'] = pd.DatetimeIndex(df_daily_new['datetime']).day_name()
df_daily_new['hour'] = df_daily_new['datetime'].dt.hour
df_daily_new['minute'] = df_daily_new['datetime'].dt.minute
df_daily_new['dayOfWeek'] = df_daily_new['datetime'].dt.dayofweek
df_daily_new['weekend'] = df_daily_new['dayOfWeek'].apply(lambda x: 1 if x >= 5 else 0)
df_daily_new['time'] = df_daily_new['datetime'].dt.time 
df_daily_new['time_Str'] = df_daily_new['datetime'].dt.time.astype(str)   
df_daily_new['dayMonth'] = df_daily_new["datetime"].dt.strftime('%m-%d')
df_daily_new['hourMinute'] = df_daily_new["datetime"].dt.strftime('%H:%M')


In [ ]:
bins = [0,4,8,12,16,20,24]
labels = ['Late Night', 'Early Morning','Morning','Noon','Eve','Night']
df_daily_new['session'] = pd.cut(df_daily_new['hour'], bins=bins, labels=labels, include_lowest=True)

In [ ]:
df_daily_new = df_daily_new[~df_daily_new['datetime'].duplicated(keep = 'first')]
df_daily_new.head()

In [ ]:
df_daily_new.columns

In [ ]:
#Reorder columns
# Put non-ML friendly columns first - they may get dropped when loading into models
col_order = ['datetime', 'consumption_kWh', 'date', 'year', 'dayMonth', 'time', 'time_Str', 'hourMinute', 'session', 'monthName', 'dayName', 
             'dayOfYear', 'week', 'month', 'day', 'hour', 'minute', 'dayOfWeek', 'weekend', 'workingDay']
df_daily_new =df_daily_new.reindex(columns=col_order)

Missing Data Imputation

In [ ]:
#Investigate where missing row/hours are
# Consumption null values occur in continuous period 
# Do not use simple forward fill, impute average day profile for both days
df_5min_new[df_5min_new['consumption'].isnull()]

datetime  timestamp  consumption  solar  alwaysOn  gridImport  \
179  2020-03-23 05:15:00        NaN          NaN    NaN       NaN         NaN   
600  2020-04-09 18:15:00        NaN          NaN    NaN       NaN         NaN   
1579 2020-05-20 13:15:00        NaN          NaN    NaN       NaN         NaN   
1632 2020-05-22 18:15:00        NaN          NaN    NaN       NaN         NaN   
1633 2020-05-22 19:15:00        NaN          NaN    NaN       NaN         NaN   
1634 2020-05-22 20:15:00        NaN          NaN    NaN       NaN         NaN   
1635 2020-05-22 21:15:00        NaN          NaN    NaN       NaN         NaN   
1636 2020-05-22 22:15:00        NaN          NaN    NaN       NaN         NaN   
1637 2020-05-22 23:15:00        NaN          NaN    NaN       NaN         NaN   
1638 2020-05-23 00:15:00        NaN          NaN    NaN       NaN         NaN   
1639 2020-05-23 01:15:00        NaN          NaN    NaN       NaN         NaN   
1640 2020-05-23 02:15:00        NaN          NaN    NaN       NaN         NaN   
1641 2020-05-23 03:15:00        NaN          NaN    NaN       NaN         NaN   
1642 2020-05-23 04:15:00        NaN          NaN    NaN       NaN         NaN   
1643 2020-05-23 05:15:00        NaN          NaN    NaN       NaN         NaN   
1644 2020-05-23 06:15:00        NaN          NaN    NaN       NaN         NaN   
1645 2020-05-23 07:15:00        NaN          NaN    NaN       NaN         NaN   
1646 2020-05-23 08:15:00        NaN          NaN    NaN       NaN         NaN   
1647 2020-05-23 09:15:00        NaN          NaN    NaN       NaN         NaN   
1648 2020-05-23 10:15:00        NaN          NaN    NaN       NaN         NaN   
1649 2020-05-23 11:15:00        NaN          NaN    NaN       NaN         NaN   
1650 2020-05-23 12:15:00        NaN          NaN    NaN       NaN         NaN   
1651 2020-05-23 13:15:00        NaN          NaN    NaN       NaN         NaN   
1652 2020-05-23 14:15:00        NaN          NaN    NaN       NaN         NaN   
2375 2020-06-22 17:15:00        NaN          NaN    NaN       NaN         NaN   
5400 2020-10-26 18:15:00        NaN          NaN    NaN       NaN         NaN   
6974 2020-12-31 08:15:00        NaN          NaN    NaN       NaN         NaN   
7583 2021-01-25 17:15:00        NaN          NaN    NaN       NaN         NaN   

      gridExport  selfConsumption  selfSufficiency active reactive voltages  \
179          NaN              NaN              NaN    NaN      NaN      NaN   
600          NaN              NaN              NaN    NaN      NaN      NaN   
1579         NaN              NaN              NaN    NaN      NaN      NaN   
1632         NaN              NaN              NaN    NaN      NaN      NaN   
1633         NaN              NaN              NaN    NaN      NaN      NaN   
1634         NaN              NaN              NaN    NaN      NaN      NaN   
1635         NaN              NaN              NaN    NaN      NaN      NaN   
1636         NaN              NaN              NaN    NaN      NaN      NaN   
1637         NaN              NaN              NaN    NaN      NaN      NaN   
1638         NaN              NaN              NaN    NaN      NaN      NaN   
1639         NaN              NaN              NaN    NaN      NaN      NaN   
1640         NaN              NaN              NaN    NaN      NaN      NaN   
1641         NaN              NaN              NaN    NaN      NaN      NaN   
1642         NaN              NaN              NaN    NaN      NaN      NaN   
1643         NaN              NaN              NaN    NaN      NaN      NaN   
1644         NaN              NaN              NaN    NaN      NaN      NaN   
1645         NaN              NaN              NaN    NaN      NaN      NaN   
1646         NaN              NaN              NaN    NaN      NaN      NaN   
1647         NaN              NaN              NaN    NaN      NaN      NaN   
1648         NaN              NaN              NaN    NaN      NaN      NaN   
1649   

In [ ]:
#Check out missing period
plt.plot(df_5min_new[(df_5min_new['datetime']>='2020-05-20 00:00:00') & (df_5min_new['datetime']<='2020-05-30 00:00:00')]['consumption'])

In [ ]:
#Get average Weekday and Weekend profiles for May and impute profile
# create_monthly_profiles(df, column)
profile_df = create_monthly_profiles(df_5min_new, 'consumption')
profile_df

timestamp  consumption       solar     alwaysOn  \
weekend monthName hour                                                       
0       April     0     1.586961e+12   152.842455   43.674318  1310.332364   
                  1     1.586965e+12   151.225545   46.842318  1310.248909   
                  2     1.586969e+12   154.196000   50.481318  1310.248909   
                  3     1.586972e+12   148.575409   46.877773  1310.248909   
                  4     1.586976e+12   151.931455   44.524773  1310.492727   
...                              ...          ...         ...          ...   
1       September 19    1.600276e+12   269.611125  138.398375  1706.781000   
                  20    1.600280e+12   195.710000   62.161000  1706.781000   
                  21    1.600284e+12   196.226000   63.231750  1706.781000   
                  22    1.600287e+12   207.688250   65.407625  1706.781000   
                  23    1.600291e+12   191.348625   61.764125  1699.147500   

                        gridImport  gridExport  selfConsumption  \
weekend monthName hour                                            
0       April     0     109.167591         0.0            100.0   
                  1     104.382727         0.0            100.0   
                  2     103.714227         0.0            100.0   
                  3     101.697182         0.0            100.0   
                  4     107.406091         0.0            100.0   
...                            ...         ...              ...   
1       September 19    131.212125         0.0            100.0   
                  20    133.548625         0.0            100.0   
                  21    132.993500         0.0            100.0   
                  22    142.280125         0.0            100.0   
                  23    129.584250         0.0            100.0   

                        selfSufficiency    year   dayOfYear  month       week  \
weekend monthName hour                                                          
0       April     0           28.739545  2020.0  106.681818    4.0  16.090909   
                  1           31.017273  2020.0  106.681818    4.0  16.090909   
                  2           32.561818  2020.0  106.681818    4.0  16.090909   
                  3           31.488636  2020.0  106.681818    4.0  16.090909   
                  4           29.461818  2020.0  106.681818    4.0  16.090909   
...                                 ...     ...         ...    ...        ...   
1       September 19          43.893750  2020.0  260.000000    9.0  37.500000   
                  20          33.783750  2020.0  260.000000    9.0  37.500000   
                  21          34.252500  2020.0  260.000000    9.0  37.500000   
                  22          33.755000  2020.0  260.000000    9.0  37.500000   
                  23          34.497500  2020.0  260.000000    9.0  37.500000   

                              day  minute  dayOfWeek  
weekend monthName hour                                
0       April     0     15.681818    15.0   2.045455  
                  1     15.681818    15.0   2.045455  
                  2     15.681818    15.0   2.045455  
                  3     15.681818    15.0   2.045455  
                  4     15.681818    15.0   2.045455  
...                           ...     ...        ...  
1       September 19    16.000000    15.0   5.500000  
                  20    16.000000    15.0   5.500000  
                  21    16.000000    15.0   5.500000  
                  22    16.000000    15.0   5.500000  
                  23    16.000000    15.0   5.500000  

[576 rows x 15 columns]

In [ ]:
#weather_df = pd.read_csv("/Users/nancy/Downloads/Data/Test/data/Harare_Hourly_Weather_2005-2021.csv")
weather_df = pd.read_csv("https://raw.githubusercontent.com/OmdenaAI/NeedEnergy/Task-2-Data-Wrangling/data/weather_data/Harare_Hourly_Weather_2005-2021.csv?token=AGZP7CAUIRLTZHJM6WLHASTAHD3Z2")
weather_df.head()

dt                         dt_iso  timezone city_name        lat  \
0  1104537600  2005-01-01 00:00:00 +0000 UTC      7200    Harare -17.821629   
1  1104541200  2005-01-01 01:00:00 +0000 UTC      7200    Harare -17.821629   
2  1104544800  2005-01-01 02:00:00 +0000 UTC      7200    Harare -17.821629   
3  1104548400  2005-01-01 03:00:00 +0000 UTC      7200    Harare -17.821629   
4  1104552000  2005-01-01 04:00:00 +0000 UTC      7200    Harare -17.821629   

         lon   temp  feels_like  temp_min  temp_max  pressure  sea_level  \
0  31.049226  17.07       17.85     16.06     17.40      1011        NaN   
1  31.049226  16.85       18.03     16.01     17.28      1016        NaN   
2  31.049226  16.70       18.17     15.78     17.13      1017        NaN   
3  31.049226  17.23       18.84     16.70     18.43      1018        NaN   
4  31.049226  17.57       18.82     17.00     18.47      1018        NaN   

   grnd_level  humidity  wind_speed  wind_deg  rain_1h  rain_3h  snow_1h  \
0         NaN        91        1.50        90      NaN      NaN      NaN   
1         NaN        93        1.00       100      NaN      NaN      NaN   
2         NaN        93        0.50        90      NaN      NaN      NaN   
3         NaN        92        0.50       140      NaN      NaN      NaN   
4         NaN        88        0.82        51      NaN      NaN      NaN   

   snow_3h  clouds_all  weather_id  weather_main     weather_description  \
0      NaN          96         211  Thunderstorm  proximity thunderstorm   
1      NaN           0         800         Clear            sky is clear   
2      NaN           0         800         Clear            sky is clear   
3      NaN          75         803        Clouds           broken clouds   
4      NaN          75         803        Clouds           broken clouds   

  weather_icon  
0          11n  
1          01n  
2          01n  
3          04n  
4          04d

In [ ]:
# Temperature, humidity and cloud cover are wanted variables
weather_df.columns

In [ ]:
weather_df['datetime'] = weather_df['datetime'].apply(lambda x: datetime.strptime(x,'%Y-%m-%d %H:%M:%S %z %Z').strftime('%Y-%m-%d %H:%M:%S'))
weather_df.head()

In [ ]:

# Slice datarame for 2020 onwards
weather_df = weather_df.loc[pd.to_datetime(weather_df['datetime']).dt.year >= 2020]
weather_df

In [ ]:
# Check there are no missing rows/hours
weather_start = weather_df['datetime'].min()
weather_end = weather_df['datetime'].max()
weather_full_index = pd.date_range(weather_start, weather_end, freq='H')
len(weather_full_index) == len(weather_df)

In [ ]:
weather_df.set_index('datetime', inplace = True)
weather_df = weather_df.reindex(weather_full_index)
weather_df.index

In [ ]:
numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
weather_features = weather_df.select_dtypes(include=numerics).columns 
for feature in weather_features:
    weather_df[feature] = weather_df[feature].replace(0,np.nan)
    weather_df[feature] = weather_df[feature].fillna(method = 'bfill')   
weather_df.reset_index(inplace = True)
weather_df.rename(columns = {'index':'datetime'}, inplace = True)
weather_df.head()

In [ ]:
#JMcQ - check the weather data is ok
# Negative value for cloud cover!!!
weather_df.describe()

In [ ]:
weather_df['temp'].plot()

In [ ]:
weather_df['humidity'].plot()

In [ ]:

# Error!
weather_df['clouds_all'].plot()

In [ ]:

# JMcQ
weather_df[weather_df['clouds_all']<0]

In [ ]:

# JMcQ
# Check period where cloud cover is less than 0
weather_df[(weather_df['datetime']>='2020-11-01 04:00:00') & (weather_df['datetime']<='2020-11-01 10:00:00')]

In [ ]:
# Use simple forward fill to replace negative cloud cover values
weather_df['clouds_all']=weather_df['clouds_all'].mask(weather_df['clouds_all']<0).ffill(downcast='infer')

In [ ]:

weather_df['clouds_all'].plot()

In [ ]:

# Rename and divide by 100 to make it more ML friendly
weather_df['clouds_all'] = weather_df['clouds_all']/100
weather_df.rename(columns={'clouds_all':'cloud_cover'}, inplace=True)
# Temperature in degrees C, rename with units
weather_df.rename(columns={'temp':'temp_degreeC'}, inplace=True)
# Humidity is relative humidity as a %
# Rename and divide by 100 to make it more ML friendly
weather_df['humidity'] = weather_df['humidity']/100
weather_df.rename(columns={'humidity':'rel_humidity'}, inplace=True)



#DATA QUALITY CHECK:
*Data summary:
* Raw_NeedEnergyApi_5min :
Raw_Records: 95649 with 15 columns
Cleaned(Current) :95649,

* Raw_NeedEnergyApi_hourly :
Raw_Record :7979 with 15 columns
Cleaned(Current) :7979,
* Raw_NeedEnergyApi_Daily :
Raw_Records: 335 with 15 columns
Cleaned (current):335
* Raw_NeedEnergyApi_Monthly :
Raw_records:12 with 15 columns
cleaned =12
* BASED ON THE FOLLOWING QUALITY DIMENSIONS:(ACCURACY,COMPELETENESS,VALUES FREE FROM CONTRADICTION)
THE NEEDENERGY API 5MIN,HOURLY,DAILTY,MONTHLY DATASET has zero missing values.More data required for predictive analysis,

## Feature Engineering:
More features were extracted and created from the date column feature:
* New features Generated :day,hour,month,year,day_name,
seasons(winter,autumn,spring),day sections(Night,Evening,Afternoon)

* Others:Jupyter starter files for 
Generalized Helper functions were created to generate seasons,day_sections,day,month,year,hour,quarter,were created for use for anyone willing to start-up cleaning or other task activities

* In-Progress Helper function
Public-holiday and many more for solar-iiradiance dataset.
